### ISSUE TITLE
Cannot import name 'TypeGuard' from 'typing_extensions'

### ISSUE THREAD

**Issue by asears** [bokeh/bokeh#12075]
### Software versions

Python version      :  3.8.8 (default, Apr 13 2021, 15:08:03) [MSC v.1916 64 bit (AMD64)]
IPython version     :  (not installed)
Tornado version     :  6.1
Bokeh version       :  3.0.0dev5+24.ga0cd142e9
BokehJS static path :  d:\projects\python\bokeh\bokeh\server\static
node.js version     :  v17.0.1
npm version         :  8.6.0
Operating system    :  Windows-10-10.0.19041-SP0

### Expected behavior

I was able to replicate this a couple ways:

## Databricks

1. Install bokeh using pip on latest Databricks 10.4 LTS cluster. https://docs.databricks.com/release-notes/runtime/10.4.html
2. Import bokeh dependency
3. TypeGuard error

## Local Dev

1. Forked/cloned the repo and installed locally.
2. Run the tests
3. TypeGuard error

### Observed behavior

ImportError: Error importing plugin "bokeh._testing.plugins.managed_server_loop": cannot import name 'TypeGuard' from 'typing_extensions' (d:\ProgramData\Anaconda3\lib\site-packages\typing_extensions.py)

### Example code

```Python
pip install -r requirements.txt 
bokeh info
pytest --color=yes tests/codebase
```

### Stack traceback or browser console output

```shell
pytest --color=yes tests/codebase
Traceback (most recent call last):
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 703, in import_plugin
    __import__(importspec)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\assertion\rewrite.py", line 170, in exec_module
    exec(co, module.__dict__)
  File "D:\projects\python\bokeh\bokeh\_testing\plugins\managed_server_loop.py", line 37, in <module>
    from bokeh.server.server import Server
  File "D:\projects\python\bokeh\bokeh\server\server.py", line 56, in <module>
    from ..core import properties as p
  File "D:\projects\python\bokeh\bokeh\core\properties.py", line 274, in <module>
    from .property.alias import Alias; Alias
  File "D:\projects\python\bokeh\bokeh\core\property\alias.py", line 28, in <module>
    from .bases import Property
  File "D:\projects\python\bokeh\bokeh\core\property\bases.py", line 50, in <module>
    from ..has_props import HasProps
  File "D:\projects\python\bokeh\bokeh\core\has_props.py", line 61, in <module>
    from .property.descriptors import PropertyDescriptor, UnsetValueError
  File "D:\projects\python\bokeh\bokeh\core\property\descriptors.py", line 104, in <module>
    from typing_extensions import TypeGuard
ImportError: cannot import name 'TypeGuard' from 'typing_extensions' (d:\ProgramData\Anaconda3\lib\site-packages\typing_extensions.py)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "d:\ProgramData\Anaconda3\Scripts\pytest-script.py", line 10, in <module>
    sys.exit(console_main())
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 185, in console_main
    code = main()
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 143, in main
    config = _prepareconfig(args, plugins)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 318, in _prepareconfig
    config = pluginmanager.hook.pytest_cmdline_parse(
  File "d:\ProgramData\Anaconda3\lib\site-packages\pluggy\hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "d:\ProgramData\Anaconda3\lib\site-packages\pluggy\manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "d:\ProgramData\Anaconda3\lib\site-packages\pluggy\manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "d:\ProgramData\Anaconda3\lib\site-packages\pluggy\callers.py", line 203, in _multicall
    gen.send(outcome)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\helpconfig.py", line 100, in pytest_cmdline_parse
    config: Config = outcome.get_result()
  File "d:\ProgramData\Anaconda3\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "d:\ProgramData\Anaconda3\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 1003, in pytest_cmdline_parse
    self.parse(args)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 1283, in parse
    self._preparse(args, addopts=addopts)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 1191, in _preparse
    self.hook.pytest_load_initial_conftests(
  File "d:\ProgramData\Anaconda3\lib\site-packages\pluggy\hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "d:\ProgramData\Anaconda3\lib\site-packages\pluggy\manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "d:\ProgramData\Anaconda3\lib\site-packages\pluggy\manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "d:\ProgramData\Anaconda3\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "d:\ProgramData\Anaconda3\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "d:\ProgramData\Anaconda3\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 1070, in pytest_load_initial_conftests
    self.pluginmanager._set_initial_conftests(early_config.known_args_namespace)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 500, in _set_initial_conftests
    self._try_load_conftest(anchor, namespace.importmode)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 508, in _try_load_conftest
    self._getconftestmodules(anchor, importmode)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 536, in _getconftestmodules
    mod = self._importconftest(conftestpath, importmode)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 587, in _importconftest
    self.consider_conftest(mod)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 663, in consider_conftest
    self.register(conftestmodule, name=conftestmodule.__file__)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 443, in register
    self.consider_module(plugin)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 669, in consider_module
    self._import_plugin_specs(getattr(mod, "pytest_plugins", []))
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 676, in _import_plugin_specs
    self.import_plugin(import_spec)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 705, in import_plugin
    raise ImportError(
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\config\__init__.py", line 703, in import_plugin
    __import__(importspec)
  File "d:\ProgramData\Anaconda3\lib\site-packages\_pytest\assertion\rewrite.py", line 170, in exec_module
    exec(co, module.__dict__)
  File "D:\projects\python\bokeh\bokeh\_testing\plugins\managed_server_loop.py", line 37, in <module>
    from bokeh.server.server import Server
  File "D:\projects\python\bokeh\bokeh\server\server.py", line 56, in <module>
    from ..core import properties as p
  File "D:\projects\python\bokeh\bokeh\core\properties.py", line 274, in <module>
    from .property.alias import Alias; Alias
  File "D:\projects\python\bokeh\bokeh\core\property\alias.py", line 28, in <module>
    from .bases import Property
  File "D:\projects\python\bokeh\bokeh\core\property\bases.py", line 50, in <module>
    from ..has_props import HasProps
  File "D:\projects\python\bokeh\bokeh\core\has_props.py", line 61, in <module>
    from .property.descriptors import PropertyDescriptor, UnsetValueError
  File "D:\projects\python\bokeh\bokeh\core\property\descriptors.py", line 104, in <module>
    from typing_extensions import TypeGuard
ImportError: Error importing plugin "bokeh._testing.plugins.managed_server_loop": cannot import name 'TypeGuard' from 'typing_extensions' (d:\ProgramData\Anaconda3\lib\site-packages\typing_extensions.py)
```

```shell
pip freeze
-e git+https://github.com/asears/bokeh.git@a0cd142e9d7c92d86656c3986ec29e5f17b9e223#egg=bokeh
flake8==4.0.1
Jinja2==3.1.1
MarkupSafe==2.1.1
mccabe==0.6.1
numpy==1.22.3
packaging==21.3
Pillow==9.1.0
pycodestyle==2.8.0
pyflakes==2.4.0
pyparsing==3.0.7
PyYAML==6.0
tornado==6.1
typing-extensions==4.1.1
xyzservices==2022.3.0
```

Seeing this in community.

https://stackoverflow.com/questions/69174965/cannot-import-name-typeguard-from-typing-extensions

**Comment by asears** [comment]
Typing-extensions has some conditional logic to expose the TypeGuard class based on Python version.  Should bokeh handle this scenario in code or remove TypeGuard if it's unstable?  

The python version with the issue.

```python
import sys
sys.version_info[:2]
```
(3, 8)

I can confirm that after upgrading to Python 10.4 I'm no longer seeing this particular error in pytest.

The commit that may have introduced the issue is #11959.  TypeGuard is only in descriptors.py, showing.py, serialization.py

![image](https://user-images.githubusercontent.com/7253192/161430522-6fbf3b42-d93d-4433-8245-9bf6c0f985a9.png)

The TypeGuard code
https://github.com/python/typing/blob/master/typing_extensions/src/typing_extensions.py

It is then unsafe to import TypeGuard as the class isn't exposed in some scenarios?

For backwards/forwards compatibility, might be best to first try to import TypeGuard from typing then use typing_extensions, then fall back to no TypeGuard class available scenario.  Yay python types.

There isn't very much use of typeguard across Bokeh yet so could also remove until it's fully working, unless there's a simple workaround besides switching python version.

**Comment by bryevdv** [comment]
Your version of `typing_extensions` is to old. We don't need to remove anything, but evidently we need to set a floor version on `typing_extensions`.

**Comment by mattpap** [comment]
As of `typing_extensions` 4.1.1 (which is used in this bug report), `TypeGuard` is exported for all relevant Python versions (see `__all__`). Also, we already have a lower bound (3.10) that includes `TypeGuard`. I'm guessing this is an environment issue.

**Comment by bryevdv** [comment]
I think so, now that I am at a computer, I cannot reproduce any issue with Python 3.8.8 and `typing_extensions`:
```
Downloading and Extracting Packages
sqlite-3.38.2        | 1.2 MB    | ##################################################################################################################### | 100%
ca-certificates-2022 | 118 KB    | ##################################################################################################################### | 100%
python-3.8.8         | 20.9 MB   | ##################################################################################################################### | 100%
typing_extensions-4. | 28 KB     | ##################################################################################################################### | 100%
openssl-1.1.1n       | 2.2 MB    | ##################################################################################################################### | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate py388
#
# To deactivate an active environment, use
#
#     $ conda deactivate


~/work/bokeh bryanv/misc_sampledata 1m 22s
base ❯ conda activate py388

~/work/bokeh bryanv/misc_sampledata
py388 ❯ python
Python 3.8.8 (default, Apr 13 2021, 12:59:45)
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing_extensions import TypeGuard
>>>
```
@asears we would need more detailed steps to reproduce before considering any action.

**Comment by asears** [comment]
Thanks for looking into this so quickly. For sure it looks like an older typing-extensions causing the issue in base env.  It's just random "luck" that I caught this in an older environment on one of my machines.

```shell
ImportError: cannot import name 'TypeGuard' from 'typing_extensions' (d:\ProgramData\Anaconda3\Lib\site-packages\typing_extensions.py)
```

typing_extensions-3.7.4.3.dist-info is what I see in the site-packages folder.

This is probably the Python 3.8.8 problem env which doesn't have TypeGuard class.

When I run a pip freeze I see typing-extensions==4.1.1 so that's interesting...

I see some related info here.
https://github.com/psf/black/issues/1928
https://github.com/tensorflow/tensorflow/issues/51743
https://github.com/numpy/numpy/pull/18285/files

I'll test again with latest Databricks next week.  Databricks says in their documentation they are running Numpy 1.20 in latest cluster but the typing-extensions version was bumped in 1.21 and further changes in 1.22.  Last week I installed bokeh (which failed on the TypeGuard error when importing) and then added typing-extensions to try and fix (which didn't fix it).

So I guess it might be a bit sketchy to use TypeGuard feature, at least when dealing with older numpy or tensorflow clobbering shared dependencies.  May want to put in some additional conditional import / usage check just in case since it's really just for static checks and not really fundamental to Bokeh?  

I guess to repro this you would need to setup bokeh then install the version of numpy or tensorflow used in Databricks 10.4 LTS clusters.

It could also be something conda env is resolving downwards to based on shared dependency.

Unfortunately I don't have an easy repro other than clobbering typing-extensions with a lower version which is obviously wrong but seems to be happening somehow.

**Comment by bryevdv** [comment]
> I'll test again with latest Databricks next week. Databricks says in their documentation they are running Numpy 1.20 in latest cluster but the typing-extensions version was bumped in 1.21 and further changes in 1.22. Last week I installed bokeh (which failed on the TypeGuard error when importing) and then added typing-extensions to try and fix (which didn't fix it).

I'm not sure how anything about numpy is relevant. Adding numpy (1.20) to the previous env does not cause any issues:

```
~/work/bokeh bryanv/misc_sampledata 15s
py388 ❯ python
Python 3.8.8 (default, Apr 13 2021, 12:59:45)
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np

>>> np.__version__
'1.20.3'
>>> from typing_extensions import TypeGuard
>>>
```

> So I guess it might be a bit sketchy to use TypeGuard feature

It's really not. No Python package can be expected to work unless all of its dependencies are correctly installed. Why would anything be different in this situation? `typing_extensions` is maintained and released by the Python project itself. We stipulate a sufficient version of `typing_extensions` for the features we use, which are only backported features that have been officially accepted in a PEP and have made it into some Python version.  

I am not sure what is going awry in your specific env, and I definitely am interested to help you find a particular solution to your particular problem. But I *do* think it is a particular problem to your environment, and not something that requires action on our end.

**Comment by bryevdv** [comment]
@mattpap I would be 👍 to close this with `noaction`

**Comment by AliHassanRevel** [comment]
> 

By installing typing-extension==4.1.1 the problem was completely solved in python==3.7. Thanks mate
